include/emp/web/Color.hpp

```diff
@@ -0,0 +1,366 @@
+    // Convert to lowercase.
+    std::transform(str.begin(), str.end(), str.begin(), ::tolower);
+
+    for (const auto& namedColor : color_impl::namedColors) {
+        if (str == namedColor.name) {
+            *this = namedColor.color;
+            return;
+        }
+    }
+
+    ParseColor(str, css_str);
+    
+    return;
+
+  }
+  
+  // Handle all color detection and parsing from the string
+  Color & Color::ParseColor(const std::string str, const std::string& css_str) {
+      // #abc and #abc123 syntax.
+    if (DetectABC(str))
+        return ParseABC(str, css_str);
+
+    size_t op = str.find_first_of('('), ep = str.find_first_of(')');
+    if (op != std::string::npos && ep + 1 == str.length()) {
+        const std::string fname = str.substr(0, op);
+        const std::vector<std::string> params = color_impl::split(str.substr(op + 1, ep - (op + 1)), ',');
+
+        float alpha = 1.0f;
+        if (fname == "rgba" || fname == "rgb") {
+            return ParseRGB(str, css_str, fname, params, alpha);
+
+        } else if (fname == "hsla" || fname == "hsl") {
+            return ParseHSL(str, css_str, fname, params, alpha);
+        }
+        return *this;
+    }
+
+    emp_assert( false, css_str );
+    return *this;
+  }
+
+  bool Color::DetectABC(const std::string str) {
+      if (str.length() && str.front() == '#')
+          return true;
+      return false;
+  }
+  Color & Color::ParseABC(const std::string str, const std::string& css_str){
+      if (str.length() == 4) {
+            int64_t iv = color_impl::parseInt(str.substr(1), 16);  // TODO(deanm): Stricter parsing.
+            if (!(iv >= 0 && iv <= 0xfff)) {
+                emp_assert( false, css_str );
+                return *this;
+            } else {
+                *this = {
+                    static_cast<uint8_t>(((iv & 0xf00) >> 4) | ((iv & 0xf00) >> 8)),
+                    static_cast<uint8_t>((iv & 0xf0) | ((iv & 0xf0) >> 4)),
+                    static_cast<uint8_t>((iv & 0xf) | ((iv & 0xf) << 4)),
+                    1
+                };
+                return *this;
+            }
+        } else if (str.length() == 7) {
+            int64_t iv = color_impl::parseInt(str.substr(1), 16);  // TODO(deanm): Stricter parsing.
+            if (!(iv >= 0 && iv <= 0xffffff)) {
+              emp_assert( false, css_str );
+              return *this;  // Covers NaN.
+            } else {
+                *this = {
+                    static_cast<uint8_t>((iv & 0xff0000) >> 16),
+                    static_cast<uint8_t>((iv & 0xff00) >> 8),
+                    static_cast<uint8_t>(iv & 0xff),
+                    1
+                };
+                return *this;
+            }
+        }
+
+        emp_assert( false, css_str );
+        return *this;
+  }
+  
+  Color & Color::ParseRGB(const std::string str, const std::string& css_str, const std::string fname,
+        const std::vector<std::string> params, float alpha){
+      if (fname == "rgba") {
```

Comment:
instead of passing the fname into this function, let's write two functions one called parse rgba and one called parse rgb